### PR TITLE
'null' is no longer used for null values - empty string instead

### PR DESCRIPTION
**Pull request by ee-usgs**
Before making a pull request
----------------------------

- [x] Make sure all tests run
- [x] Update the changelog appropriately

'null' String is no longer used for null values - empty string is used instead
-----------
Ref: https://internal.cida.usgs.gov/jira/browse/IOW-683
* Improved and more deterministic testing for output values
* RdbWriter has more readable row count methods
* Added test specifically for null/empty values
* Updated readme w/ additional details on testing (missing a db)

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [x] If appropriate, put the link to the PR in the JIRA ticket
- [x] Assign someone to review unless the change is trivial

### CODE REVIEW

File: src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java [hunk 1]
```diff
@@ -38,10 +38,10 @@ public RdbWriter(Writer destination) {
 		initRows();
 	}
 
-	public long getHeaderRows() {
+	public long getHeaderRowCount() {
 		return headerLineCount;
 	}
-	public long getDataRows() {
+	public long getDataRowCount() {
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java:44]: These method names kept tripping me up, so I changed them.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java [hunk 2]
```diff
@@ -130,16 +130,17 @@ protected RdbWriter writeRow(List<String> columns) {
...
+
 		if (value == null) {
 			value = "";
 		}
 		if (value.length() > length) {
-			trimmedValue = value.substring(0, length);
+			value = value.substring(0, length);
 		}
-		return trimmedValue;
+		return value;
```
Comment (ee-usgs) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java:144]: This tricky bit of code actually let null values thru (even though it didn't look like it)

File: src/test/java/gov/usgs/wma/waterdata/groundwater/RdbWriterTest.java [hunk 3]
```diff
@@ -100,40 +106,101 @@ void testHeaderWriten() {
...
+		assertEquals("USGS", rdbWriter.getLastValueFor("lev_agency_cd"));
+		assertEquals("S", rdbWriter.getLastValueFor("lev_meth_cd"));
+		assertEquals("T", rdbWriter.getLastValueFor("lev_age_cd"));
+		assertEquals("20070501", rdbWriter.getLastValueFor("lev_dt"));
+		assertEquals("1200", rdbWriter.getLastValueFor("lev_tm"));
+		assertEquals("01-MAY-2007 12:00:00", rdbWriter.getLastValueFor("lev_dtm"));
+		assertEquals("07-MAY-2007 18:30:47", rdbWriter.getLastValueFor("lev_md"));
+		assertEquals("UTC", rdbWriter.getLastValueFor("lev_tz_cd"));
+		assertEquals("D", rdbWriter.getLastValueFor("lev_dt_acy_cd"));
+		assertEquals("30210", rdbWriter.getLastValueFor("parameter_code"));
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/wma/waterdata/groundwater/RdbWriterTest.java:202]: I needed to be more specific when investigating this, so I expanded the mock rdbWriter to be able to match up the column names and values.

File: src/test/java/gov/usgs/wma/waterdata/groundwater/RdbWriterTest.java [hunk 4]
```diff
@@ -229,28 +294,78 @@ public Writer append(CharSequence csq) throws IOException {
...
+				e.printStackTrace(); //ignore
 			}
+		}
+		@Override
+		public RdbWriter writeHeader() {
+			super.writeHeader();
+			flush();
+			headerRaw = out.toString();
 			return this;
 		}
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/wma/waterdata/groundwater/RdbWriterTest.java:318]: This method no longer needs to be explicitly called from the test, so I internalized it.

File: src/test/java/gov/usgs/wma/waterdata/groundwater/RdbWriterTest.java [hunk 5]
```diff
@@ -229,28 +294,78 @@ public Writer append(CharSequence csq) throws IOException {
...
+			super.writeRow(columns);
+
+			flush();
+
+			//Append the last line to our running list of dataRows
+			String[] lines = out.toString().split("(\\r\\n|\\r|\\n)", -1);
+			dataRows.add(lines[lines.length - 2]);  //A trailing newline is always added, so '-1' would be an empty row
+
 			return this;
 		}
```
Comment (ee-usgs) [inline src/test/java/gov/usgs/wma/waterdata/groundwater/RdbWriterTest.java:331]: 'close()' is no longer needed.

File: src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java (hunk 2 above)
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java:144]: Good catch

File: src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java [hunk 6]
```diff
@@ -38,10 +38,10 @@ public RdbWriter(Writer destination) {
 		initRows();
 	}
 
-	public long getHeaderRows() {
+	public long getHeaderRowCount() {
```
Comment (ssoper-usgs) [inline src/main/java/gov/usgs/wma/waterdata/groundwater/RdbWriter.java:41]: Good name change here, the previous name made me think we were maybe going to get a list of string values or pojo property values or something.

**Review by ssoper-usgs — APPROVED** [review]

**Comment by ee-usgs** [comment]
@duselmann I'm merging this now after Sam's approval.  Your markups are welcome.

**Review by duselmann — COMMENTED** [review]
looks good to me.

**Pull request merged**